wire-runtime/src/main/java/com/squareup/wire/Message.java

```diff
@@ -143,41 +83,57 @@ protected void setBuilder(Builder builder) {
     }
     if (list == Collections.emptyList()) {
       return list;
-    } else if (list instanceof MessageAdapter.ImmutableList) {
-      return list;
     }
     return Collections.unmodifiableList(new ArrayList<T>(list));
   }
 
-  /**
-   * Returns the enumerated value tagged with the given integer value for the
-   * given enum class. If no enum value in the given class is initialized
-   * with the given integer tag value, an exception will be thrown.
-   *
-   * @param <E> the enum class type
-   */
-  public static <E extends Enum & ProtoEnum> E enumFromInt(Class<E> enumClass, int value) {
-    EnumAdapter<E> adapter = WIRE.enumAdapter(enumClass);
-    return adapter.fromInt(value);
+  protected abstract void visitFields(Visitor visitor);
+
+  final int serializedSize() {
+    int serializedSize = this.serializedSize;
+    if (serializedSize == -1) {
+      SerializedSizeVisitor visitor = new SerializedSizeVisitor();
+      visitFields(visitor);
+      serializedSize = this.serializedSize = visitor.size;
+    }
+    return serializedSize;
   }
 
-  void writeUnknownFieldMap(WireOutput output) throws IOException {
-    if (unknownFields != null) {
-      unknownFields.write(output);
+  final void write(ProtoWriter writer) throws IOException {
+    try {
+      visitFields(writer.visitor);
+    } catch (WriteIOException e) {
+      throw e.cause;
     }
   }
 
-  int getUnknownFieldsSerializedSize() {
-    return unknownFields == null ? 0 : unknownFields.getSerializedSize();
+  @Override public final String toString() {
+    ToStringVisitor visitor = new ToStringVisitor();
+    visitFields(visitor);
+    return visitor.finish(this);
   }
 
-  protected static boolean equals(Object a, Object b) {
-    return a == b || (a != null && a.equals(b));
+  @Override public final int hashCode() {
+    int hashCode = this.hashCode;
+    if (hashCode == 0) {
+      HashCodeVisitor visitor = new HashCodeVisitor();
+      visitFields(visitor);
+      hashCode = this.hashCode = visitor.result;
+    }
+    return hashCode;
   }
 
-  @SuppressWarnings("unchecked")
-  @Override public String toString() {
-    return WIRE.messageAdapter((Class<Message>) getClass()).toString(this);
+  @Override public final boolean equals(Object o) {
+    if (o == this) return true;
+    if (o == null || o.getClass() != getClass()) return false;
+
+    EqualsVisitor thisVisitor = new EqualsVisitor();
+    visitFields(thisVisitor);
+
+    EqualsVisitor otherVisitor = new EqualsVisitor();
+    ((Message<?>) o).visitFields(otherVisitor);
+
+    return thisVisitor.isEqualTo(otherVisitor);
```

Comment:
While this is truly clever, it's pretty inefficient since there is no possibility of early exit. On the other hand, I would imaging that most calls to `equals` happen only when there is a hash code collision, which is extremely rare. We could always go back to codegen for `equals` if this is a problem.